staging/src/k8s.io/api/coordination/v1/types.go

```diff
@@ -20,6 +20,22 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+type CoordinatedStrategy string
+
+// CoordinatedLeaseStrategy defines the strategy for picking the leader for coordinated leader election.
+const (
+	// OldestEmulationVersion picks the oldest LeaseCandidate, where "oldest" is defined as follows
+	// 1) Select the candidate with the emulation version
+	// 2) If multiple candidates have the same emulation version, select the candidate(s) with the lowest binary version. (Note that binary version must be greater or equal to emulation version)
+	// 3) If multiple candidates have the same binary version, select the candidate with the oldest creationTimestamp.
+	// If a candidate specifies a binary version but not a emulation version, the binary version will be assumed. If a candidate specifies an emulation version but not a binary version, the emulation version will be assumed.
```

Comment:
Now that Strategy is defined in the LeaseCandidate I think it might make more sense to explicitly indicate no defaulting since we can conditionally validate these fields during the Create/Update of a LeaseCandidate.

Reply:
Not sure I follow.  We could say (for example and discussion)

* `binVer` and `compatVer` are required when `strategy` is "Oldest"
* if exactly one of `binVer` or `compatVer` is specified, the unspecified field is defaulted from the specified field

This is clean but makes more verbose YAML.

Or we could say:

* `binVer` and `compatVer` are strictly optional
* When `strategy` is "Oldest", one or both of `binVer` and `compatVer` must be specified, or the candidate will not be considered
* if exactly one of `binVer` or `compatVer` is specified, the unspecified field is implied from the specified field, but not defaulted

Now we a) need to indicate somewhere when a LeaseCandidate is being ignored (which we probably need to do anyway, in case of incompatible strategies); and b) need clients to implement the implicit value.

Or we could say:

* `binVer` and `compatVer` are strictly optional, but when `strategy` is "Oldest", one or both of `binVer` and `compatVer` must be specified
* if exactly one of `binVer` or `compatVer` is specified, the unspecified field is implied from the specified field, but not defaulted

This dodges the need to indicate somewhere when a LeaseCandidate is being ignored (but we still need to do that), and clients still need to implement the implicit value.

I could go in any direction on this, really - are there other tradeoffs I am not considering or weighinig approporiately?  In k8s, we USUALLY opt for explicit defaulting, but that's more to capture API-version-specific semantics (e.g. default is foo in v1 but bar in v2).  So it doesn't matter SO much here.

Reply:
My preference:
- binVer and compatVer are strictly optional
- When strategy is "Oldest", **BOTH** `binVer` and `compatVer` must be specified, or the candidate will not be considered
- No defaulting or implications, just error out on missing fields

I'd like users to be explicit with `binVer` and `compatVer` so we don't get any surprises. imo if a component only has a binaryVersion, logically `compatVer == binVer`. I think its difficult to express this logic with defaulting. 

If we default `compatVer` to `binaryVer` or vice versa, what if users want to update `binVer`? Seeing `compatVer` stuck at a particular number is not ideal, as the component will have `compatVer != binVer` after the update. If another candidate joins at the latest binVer, it will have precedent with a higher `compatVer`. 

If we default the missing field to something like MAXVERSION or MINVERSION, it preserves behavior but feels inaccurate. We would also have to deal with edge cases where users may want to start using and setting the missing field and ensuring compatibility there. 

Can we simply ask users to be explicit and declare both? If their component only has one version, create the LeaseCandidate with both fields but set them to be equal.

Reply:
Updates are a good argument for not doing explicit defaults.  Thanks.

> Can we simply ask users to be explicit and declare both? 

We can.  That's fine if that is the semantic you think is best.  Please be very clear in the doc-comments around these fields that BOTH need to be set.

Now on to error handling.

When we try to do an election, we need to be able to report things like "Strategy was 'Oldest', but candidate X did not set the binaryVersion or compatVersion field".  That has to be a runtime thing, since the CLE-chosen Lease strategy might not be the one that any given candidate preferred.

When we API-admit a LeaseCandidate, is it a validation error if the user specifies one but not both?  IOW:

```
if strategy == Oldest {
    if spec.CompatVersion == "" {
        // field.Required(...)
    }
    if spec.BinaryVersion == "" {
        // field.Required(...)
    }
}
```

Or do we JUST rely on the runtime error?

Reply:
This got me thinking about whether []preferredStrategies should be preferred or supported strategies? If candidate X only supports strategy A, and candidate Y comes in and trumps strategy A with preferredStrategies [B, A], we do not validate strategy B fields for candidate X. It "could" fill in those fields but we do no validation for it. Behavior seems quite conditional since we may get partial fields, have to do the full validation again, and validation between the controller and resource could drift. Can we just exclude it from leader selection on strategy B?

Reply:
I don't think it's "supported".  That implies a negotiation or as you said, exclusion.  The specific case @sttts described was more like this:

As a component author, I chose `strategy=OldestCompatVersion`.  All the running replicas support it in their LeaseCandidates.  Over time I decide that I really want the new `OldestCompatVersionNewestBinary` strategy.  It uses the same information, but makes a different election decision.  I can roll this out in my next release, saying I want "OldestCompatVersionNewestBinary" and if the strategy is currently "OldestCompatVersion", I win.  It's implied that I know what I am doing.

Or I could decide that I want to switch to some other strategy, so I do 1 release which populates the new information but uses the old strat, and then the next release switches the strategy.  Again,  I know what I am doing.

If we really find ourselves with 2 candidates who ask for differing and non-overlapping strategies, we have a real problem.  We need to spec what happens, but it doesn't need to be "useful".